mithril-aggregator/src/tools/file_archiver/appender.rs

```diff
@@ -157,6 +167,96 @@ impl TarAppender for AppenderEntries {
     }
 }
 
+/// Append data to the archive.
+pub struct AppenderData {
+    /// Location of the file in the archive where the data will be appended.
+    location_in_archive: PathBuf,
+    /// Byte array of the data to append.
+    bytes: Vec<u8>,
+}
+
+impl AppenderData {
+    /// Create a new instance of `AppenderData` from an object that will be serialized to JSON.
+    pub fn from_json<T: Serialize + Send>(
+        location_in_archive: PathBuf,
+        object: &T,
+    ) -> StdResult<Self> {
+        let json_bytes = serde_json::to_vec(object).with_context(|| {
+            format!(
+                "Can not serialize JSON to file in archive: {:?}",
+                location_in_archive.display()
+            )
+        })?;
+
+        Ok(Self::from_raw_bytes(location_in_archive, json_bytes))
+    }
+
+    /// Create a new instance of `AppenderData` from a byte array.
+    pub fn from_raw_bytes(location_in_archive: PathBuf, bytes: Vec<u8>) -> Self {
+        Self {
+            location_in_archive,
+            bytes,
+        }
+    }
+}
+
+impl TarAppender for AppenderData {
+    fn append<T: Write>(&self, tar: &mut tar::Builder<T>) -> StdResult<()> {
+        let mut header = tar::Header::new_gnu();
+        header.set_size(self.bytes.len() as u64);
+        header.set_mode(READ_WRITE_PERMISSION);
+        header.set_mtime(chrono::Utc::now().timestamp() as u64);
+        header.set_cksum();
+
+        tar.append_data(
+            &mut header,
+            &self.location_in_archive,
+            self.bytes.as_slice(),
+        )
+        .with_context(|| {
+            format!(
+                "Can not add file: '{}' to the archive",
+                self.location_in_archive.display()
+            )
+        })?;
+
+        Ok(())
+    }
+
+    fn compute_uncompressed_data_size(&self) -> StdResult<u64> {
+        Ok(self.bytes.len() as u64)
+    }
+}
+
+/// Chain multiple `TarAppender` instances together.
+pub struct ChainAppender<A1, A2> {
+    appender_left: A1,
+    appender_right: A2,
+}
```

Comment:
Maybe this would be more readable:
```suggestion
pub struct ChainAppender<L, R> {
    appender_left: L,
    appender_right: R,
}
```